### PR TITLE
[worker, trainer, recipe] feat: add FP16 training and inference support

### PR DESCRIPTION
**Pull request by Xuekai-Zhu**
This PR adds FP16 (float16 precision) training support to verl. The implementation includes:

 | Component | Precision |
|-----------|-----------|
| **Training (Actor)** | float16 |
| **Training (Ref)** | float16 |
| **Inference (vLLM Rollout)** | float16 |
| **Gradient Scaler** | ShardedGradScaler (enabled) |

- Inspired by [Precision-RL Patch](https://github.com/sail-sg/Precision-RL/blob/main/verl_fp16.patch)￼
- Reference Paper: [Defeating the Training–Inference Mismatch via FP16](https://arxiv.org/pdf/2510.26788)￼

A new script demonstrates end-to-end FP16 training:

File: recipe/flowrl/run_flowrl_qwen2.5_7b_fp16.sh
This script launches FlowRL with the appropriate precision overrides:
```
    actor_rollout_ref.actor.dtype=float16 \
    actor_rollout_ref.ref.dtype=float16 \
    actor_rollout_ref.rollout.dtype=float16 \

```

### Implementation Details :
1. `verl/workers/actor/dp_actor.py`
```diff
+ # Add ShardedGradScaler for fp16
+ if self.config.dtype == "float16":
+     from torch.distributed.fsdp.sharded_grad_scaler import ShardedGradScaler
+     self.scaler = ShardedGradScaler(growth_interval=400)
+ else:
+     self.scaler = None

+ # Use scaler in backward
+ if self.scaler is not None:
+     self.scaler.unscale_(self.actor_optimizer)

+ # Use scaler in optimizer step
+ if self.scaler is not None:
+     self.scaler.step(self.actor_optimizer)
+     self.scaler.update()
```

2. `recipe/flowrl/flowrl_fsdp_worker.py`
```diff
- param_dtype = torch.bfloat16
+ param_dtype = PrecisionType.to_dtype(self.config.actor.get("dtype", "float16"))

+ vllm_dtype = PrecisionType.to_dtype(self.config.rollout.dtype)
- torch_dtype = torch.float32 if self._is_actor else torch.bfloat16
+ torch_dtype = torch.float32 if self._is_actor else vllm_dtype
```

3. `recipe/flowrl/flowrl_actor.py`
```diff
- with torch.autocast(device_type=self.device_name, dtype=torch.bfloat16):
+ from verl.utils.torch_dtypes import PrecisionType
+ torch_dtype = PrecisionType.to_dtype(self.config.dtype)
+ with torch.autocast(device_type=self.device_name, dtype=torch_dtype):

- loss.backward()
+ if self.scaler is not None:
+     self.scaler.scale(loss).backward()
+ else:
+     loss.backward()
```

### CODE REVIEW

File: verl/trainer/config/rollout/rollout.yaml [hunk 1]
```diff
@@ -26,7 +26,7 @@ response_length: ${oc.select:data.max_response_length,512}
 
 # for vllm rollout
 # Rollout model parameters type. Align with actor model's FSDP/Megatron type.
-dtype: bfloat16
+dtype: float16
```
Comment (ISEEKYAN) [inline verl/trainer/config/rollout/rollout.yaml]: we would better not change the default bf16 dtype, instead we can update an example script with overrode fp16 dtype.

**Comment by PeterSH6** [comment]
Same as @ISEEKYAN , bf16 should be kept.

More comprehensive tests on fp16 for RL training should be verified.

**Comment by Xuekai-Zhu** [comment]
Yes, I agree. @ISEEKYAN @PeterSH6 
I’ve kept the default bf16 dtype unchanged, and added a new example script in the flowrl recipe that enables FP16 through configuration overrides.

recipe/flowrl/run_flowrl_qwen2.5_7b_fp16.sh
```
    actor_rollout_ref.actor.dtype=float16 \
    actor_rollout_ref.ref.dtype=float16 \
    actor_rollout_ref.rollout.dtype=float16 \
    actor_rollout_ref.actor.use_amp=True \

```

This way, users can experiment with FP16 without affecting the existing default setup.
Also, I’m currently running FP16 training experiments to further verify its stability and performance on RL workloads.
W B Chart 2025_11_6 19_09_05

**Comment by vermouth1992** [comment]
> Yes, I agree. @ISEEKYAN @PeterSH6 I’ve kept the default bf16 dtype unchanged, and added a new example script in the flowrl recipe that enables FP16 through configuration overrides.
> 
> recipe/flowrl/run_flowrl_qwen2.5_7b_fp16.sh
> 
> ```
>     actor_rollout_ref.actor.dtype=float16 \
>     actor_rollout_ref.ref.dtype=float16 \
>     actor_rollout_ref.rollout.dtype=float16 \
>     actor_rollout_ref.actor.use_amp=True \
> ```
> 
> This way, users can experiment with FP16 without affecting the existing default setup. Also, I’m currently running FP16 training experiments to further verify its stability and performance on RL workloads. W B Chart 2025_11_6 19_09_05

Could you also show the eval score? Thanks!

**Comment by Xuekai-Zhu** [comment]
> > Yes, I agree. @ISEEKYAN @PeterSH6 I’ve kept the default bf16 dtype unchanged, and added a new example script in the flowrl recipe that enables FP16 through configuration overrides.
> > recipe/flowrl/run_flowrl_qwen2.5_7b_fp16.sh
> > ```
> >     actor_rollout_ref.actor.dtype=float16 \
> >     actor_rollout_ref.ref.dtype=float16 \
> >     actor_rollout_ref.rollout.dtype=float16 \
> >     actor_rollout_ref.actor.use_amp=True \
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This way, users can experiment with FP16 without affecting the existing default setup. Also, I’m currently running FP16 training experiments to further verify its stability and performance on RL workloads. W B Chart 2025_11_6 19_09_05
> 
> Could you also show the eval score? Thanks!


Absolutely! Here are the current online eval results; I’ll update more later.
W B Chart 2025_11_6 21_13_06

File: verl/workers/actor/dp_actor.py [hunk 2]
```diff
@@ -82,6 +83,13 @@ def __init__(self, config: ActorConfig, actor_module: nn.Module, actor_optimizer
             else entropy_from_logits
         )
         self.device_name = get_device_name()
+        assert self.config.dtype in ["float16", "float32", "bfloat16"]
+        if self.config.dtype == "float16":
+            from torch.distributed.fsdp.sharded_grad_scaler import ShardedGradScaler
```
Comment (wuxibin89) [inline verl/workers/actor/dp_actor.py:88]: Does `ShardedGradScaler` works for both fsdp and fsdp2?

File: verl/workers/actor/dp_actor.py [hunk 3]
```diff
@@ -274,7 +283,8 @@ def _forward_micro_batch(
 
     def _optimizer_step(self):
         assert self.config.grad_clip is not None
-
+        if self.scaler is not None:
+            self.scaler.unscale_(self.actor_optimizer)
```
Comment (wuxibin89) [inline verl/workers/actor/dp_actor.py:286]: Just curious
1. why we need `unscale_` here?
2. why we don't scale actor loss by `self.scalar(loss).backward()`?

File: verl/workers/actor/dp_actor.py (hunk 2 above)
Comment (QPHutu) [reply verl/workers/actor/dp_actor.py:88]: yes,  it works for both fsdp and fsdp2.

File: verl/workers/actor/dp_actor.py (hunk 3 above)
Comment (QPHutu) [reply verl/workers/actor/dp_actor.py:286]: 1. the `unscale_` is for `clip_grad_norm`, otherwise the grad_norm would be scaled. Reference doc: https://docs.pytorch.org/docs/stable/notes/amp_examples.html#gradient-clipping
2. yes we do need to scale loss by `self.scaler(loss).backward()`. it was already updated

**Comment by wuxibin89** [comment]
@Xuekai-Zhu Is this ready for merge? There's some ci failed.

**Comment by Xuekai-Zhu** [comment]
> @Xuekai-Zhu Is this ready for merge? There's some ci failed.

@wuxibin89  I’ll work on fixing these CI issues today !  After the fixes are ready, I’ll reach out to you again for another review. Thanks again!

File: verl/workers/fsdp_workers.py [hunk 4]
```diff
@@ -310,7 +310,13 @@ def _build_model_optimizer(
 
         torch_dtype = fsdp_config.get("model_dtype", None)
         if torch_dtype is None:
```
Comment (wuxibin89) [inline verl/workers/fsdp_workers.py]: There's 2 question here:
1. `fsdp_config.get("model_dtype", None)` always return fp32 here, so we never enter this branch.
2. `torch_dtype` only affects `AutoModel.from_pretrained`, while FSDP mixed precision still use bf16 as param_dtype
https://github.com/volcengine/verl/blob/main/verl/workers/fsdp_workers.py#L1346-L1348

**Review by wuxibin89 — APPROVED** [review]
